source/proxy_connection.c

```diff
@@ -0,0 +1,316 @@
+    int error_code,
+    void *user_data) {
+    struct aws_http_proxy_user_data *proxy_ud = user_data;
+
+    proxy_ud->original_on_shutdown(connection, error_code, proxy_ud->original_user_data);
+
+    aws_http_proxy_user_data_destroy(user_data);
+}
+
+#define DEFAULT_BASIC_AUTH_HEADER_VALUE_SIZE 128
+
+AWS_STATIC_STRING_FROM_LITERAL(s_proxy_authorization_header_name, "Proxy-Authorization");
+AWS_STATIC_STRING_FROM_LITERAL(s_proxy_authorization_header_basic_prefix, "Basic ");
+
+static int s_add_basic_proxy_authentication_header(
+    struct aws_http_message *request,
+    struct aws_http_proxy_user_data *proxy_user_data) {
+    struct aws_byte_buf base64_input_value;
+    AWS_ZERO_STRUCT(base64_input_value);
+
+    struct aws_byte_buf header_value;
+    AWS_ZERO_STRUCT(header_value);
+
+    int result = AWS_OP_ERR;
+
+    if (aws_byte_buf_init(
+            &base64_input_value,
+            proxy_user_data->allocator,
+            proxy_user_data->username->len + proxy_user_data->password->len + 1)) {
+        goto done;
+    }
+
+    /* First build a buffer with "username:password" in it */
+    struct aws_byte_cursor username_cursor = aws_byte_cursor_from_string(proxy_user_data->username);
+    if (aws_byte_buf_append(&base64_input_value, &username_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor colon_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(":");
+    if (aws_byte_buf_append(&base64_input_value, &colon_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor password_cursor = aws_byte_cursor_from_string(proxy_user_data->password);
+    if (aws_byte_buf_append(&base64_input_value, &password_cursor)) {
+        goto done;
+    }
+
+    struct aws_byte_cursor base64_source_cursor =
+        aws_byte_cursor_from_array(base64_input_value.buffer, base64_input_value.len);
+
+    /* Figure out how much room we need in our final header value buffer */
+    size_t required_size = 0;
+    if (aws_base64_compute_encoded_len(base64_source_cursor.len, &required_size)) {
+        goto done;
+    }
+
+    required_size += s_proxy_authorization_header_basic_prefix->len + 1;
+    if (aws_byte_buf_init(&header_value, proxy_user_data->allocator, required_size)) {
+        goto done;
+    }
+
+    /* Build the final header value by appending the authorization type and the base64 encoding string together */
+    struct aws_byte_cursor basic_prefix = aws_byte_cursor_from_string(s_proxy_authorization_header_basic_prefix);
+    if (aws_byte_buf_append_dynamic(&header_value, &basic_prefix)) {
+        goto done;
+    }
+
+    if (aws_base64_encode(&base64_source_cursor, &header_value)) {
+        goto done;
+    }
+
+    struct aws_http_header header = {.name = aws_byte_cursor_from_string(s_proxy_authorization_header_name),
+                                     .value = aws_byte_cursor_from_array(header_value.buffer, header_value.len)};
+
+    if (aws_http_message_add_header(request, header)) {
```

Comment:
Is there any chance the authorization header already exists, and we should be appending to it, with like comma separated values or anything weird like that?

Reply:
In that case you'd just use no authentication; I don't think it's worth checking for duplicates